gio/src/subclass/list_model.rs

```diff
@@ -0,0 +1,83 @@
+use crate::auto::ListModel;
+use glib::subclass::prelude::*;
+use glib::translate::*;
+use glib::Cast;
+use glib::ObjectExt;
+
+pub trait ListModelImpl: ObjectImpl {
+    fn get_item_type(&self, list_model: &Self::Type) -> glib::Type;
+    fn get_n_items(&self, list_model: &Self::Type) -> u32;
+    fn get_item(&self, list_model: &Self::Type, position: u32) -> Option<glib::Object>;
+}
+
+unsafe impl<T: ListModelImpl + glib::ObjectType> IsImplementable<T> for ListModel {
+    unsafe extern "C" fn interface_init(
+        iface: glib::ffi::gpointer,
+        _iface_data: glib::ffi::gpointer,
+    ) {
+        let list_model_iface = &mut *(iface as *mut ffi::GListModelInterface);
+
+        list_model_iface.get_item_type = Some(list_model_get_item_type::<T>);
+        list_model_iface.get_n_items = Some(list_model_get_n_items::<T>);
+        list_model_iface.get_item = Some(list_model_get_item::<T>);
+    }
+}
+
+unsafe extern "C" fn list_model_get_item_type<T: ObjectExt + ListModelImpl>(
+    list_model: *mut ffi::GListModel,
+) -> glib::ffi::GType {
+    let instance = &*(list_model as *mut T::Instance);
```

Comment:
It's the instance that needs to be `ObjectType` and on which you can call `set`/`get_qdata`. Not sure how this works at all at runtime currently :)